File: fuzz/fuzz_targets/runner/async_run.rs
#![no_main]

use libfuzzer_sys::arbitrary::Arbitrary;
use libfuzzer_sys::fuzz_target;
use openchecks::CheckMetadata;

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Arbitrary)]
struct Item {
    value: u32,
}

impl std::fmt::Display for Item {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.value)
    }
}

impl openchecks::Item for Item {
    type Value = u32;

    fn value(&self) -> Self::Value {
        self.value
    }
}

#[derive(Debug, Clone, Arbitrary)]
struct Check {
    pub title: String,
    pub description: String,
    pub hint: openchecks::CheckHint,
    pub status: openchecks::Status,
    pub fix_status: openchecks::Status,
    pub message: String,
    pub items: Option<Vec<Item>>,
    pub can_fix: bool,
    pub can_skip: bool,
    pub error: Option<openchecks::Error>,
}

impl openchecks::CheckMetadata for Check {
    fn title(&self) -> std::borrow::Cow<str> {
        std::borrow::Cow::Borrowed(&self.title)
    }

    fn description(&self) -> std::borrow::Cow<str> {
        std::borrow::Cow::Borrowed(&self.description)
    }

    fn hint(&self) -> openchecks::CheckHint {
        self.hint
    }
}

#[async_trait::async_trait]
impl openchecks::AsyncCheck for Check {
    type Item = Item;
    type Items = Vec<Item>;

    async fn async_check(&self) -> openchecks::CheckResult<Self::Item, Self::Items> {
        openchecks::CheckResult::new(
            self.status,
            &self.message,
            self.items.clone(),
            self.can_fix,
            self.can_skip,
            self.error.clone(),
        )
    }

    async fn async_auto_fix(&mut self) -> Result<(), openchecks::Error> {
        match &self.error {
            Some(error) => Err(error.clone()),
            None => {
                self.status = self.fix_status;
                Ok(())
            }
        }
    }
}

fuzz_target!(|check: Check| {
    let mut check = check;
    assert_eq!(check.title().as_ref(), &check.title);
    assert_eq!(check.description().as_ref(), &check.description);
    assert_eq!(check.hint(), check.hint);

    let rt = tokio::runtime::Builder::new_current_thread()
        .build()
        .unwrap();

    let result = rt.block_on(async { openchecks::async_run(&check).await });

    assert_eq!(result.status(), &check.status);
    assert_eq!(result.message(), &check.message);
    assert_eq!(result.items(), &check.items);
    assert_eq!(result.error(), &check.error);

    if result.status() == &openchecks::Status::SystemError {
        assert!(!result.can_fix());
        assert!(!result.can_skip());
    } else {
        assert_eq!(result.can_fix(), check.can_fix);
        assert_eq!(result.can_skip(), check.can_skip);
    }

    if result.status().has_failed() && result.can_fix() {
        let fix_result = rt.block_on(async { openchecks::async_auto_fix(&mut check).await });

        if !check.hint().contains(openchecks::CheckHint::AUTO_FIX) {
            assert_eq!(fix_result.status(), &openchecks::Status::SystemError);
            assert_eq!(fix_result.message(), "Check does not implement auto fix.");
            assert_eq!(fix_result.items(), &None);
            assert_eq!(fix_result.error(), &None);
        } else if fix_result.error().is_some() {
            assert_eq!(fix_result.status(), &openchecks::Status::SystemError);
            assert_eq!(fix_result.message(), "Error in auto fix.");
            assert_eq!(fix_result.items(), &None);
            assert_eq!(fix_result.error(), &check.error);
        } else {
            assert_eq!(fix_result.status(), &check.fix_status);
            assert_eq!(fix_result.message(), &check.message);
            assert_eq!(fix_result.items(), &check.items);
            assert_eq!(fix_result.error(), &check.error);
        }

        if fix_result.status() == &openchecks::Status::SystemError {
            assert!(!fix_result.can_fix());
            assert!(!fix_result.can_skip());
        } else {
            assert_eq!(fix_result.can_fix(), check.can_fix);
            assert_eq!(fix_result.can_skip(), check.can_skip);
        }
    }
});

File: bindings/python/fuzz/async_run.py
from __future__ import annotations

import asyncio

import atheris
import hypothesis
from hypothesis import strategies

import openchecks


class Check(openchecks.AsyncBaseCheck):
    def __init__(  # noqa: PLR0913
        self,
        title: str,
        description: str,
        hint: openchecks.CheckHint,
        status: openchecks.Status,
        fix_status: openchecks.Status,
        message: str,
        items: list[openchecks.Item[int]] | None,
        can_fix: bool,  # noqa: FBT001
        can_skip: bool,  # noqa: FBT001
        error: BaseException | None,
    ) -> None:
        self._title = title
        self._description = description
        self._hint = hint
        self._status = status
        self._fix_status = fix_status
        self._message = message
        self._items = items
        self._can_fix = can_fix
        self._can_skip = can_skip
        self._error = error

    def title(self) -> str:
        return self._title

    def description(self) -> str:
        return self._description

    def hint(self) -> openchecks.CheckHint:
        return self._hint

    async def async_check(self) -> openchecks.CheckResult:
        return openchecks.CheckResult(
            self._status,
            self._message,
            self._items,
            self._can_fix,
            self._can_skip,
            self._error,
        )

    async def async_auto_fix(self) -> None:
        if self._error:
            raise self._error

        self._status = self._fix_status


@strategies.composite
def check_hints(draw: strategies.DrawFn) -> openchecks.CheckHint:
    hint = openchecks.CheckHint.NONE

    if draw(strategies.booleans()):
        hint |= openchecks.CheckHint.AUTO_FIX

    return hint


@hypothesis.given(
    check=strategies.builds(
        Check,
        title=strategies.text(),
        description=strategies.text(),
        hint=check_hints(),
        status=strategies.sampled_from(
            [
                openchecks.Status.Pending,
                openchecks.Status.Skipped,
                openchecks.Status.Passed,
                openchecks.Status.Warning,
                openchecks.Status.Failed,
                openchecks.Status.SystemError,
            ]
        ),
        fix_status=strategies.sampled_from(
            [
                openchecks.Status.Pending,
                openchecks.Status.Skipped,
                openchecks.Status.Passed,
                openchecks.Status.Warning,
                openchecks.Status.Failed,
                openchecks.Status.SystemError,
            ]
        ),
        message=strategies.text(),
        items=strategies.one_of(
            strategies.none(),
            strategies.lists(
                strategies.builds(
                    openchecks.Item,
                    value=strategies.integers(),
                    type_hint=strategies.one_of(strategies.none(), strategies.text()),
                )
            ),
        ),
        can_fix=strategies.booleans(),
        can_skip=strategies.booleans(),
        error=strategies.one_of(
            strategies.none(), strategies.builds(Exception, strategies.text())
        ),
    )
)
@atheris.instrument_func
def fuzz(
    check: Check,
) -> None:
    asyncio.run(_validate(check))


async def _validate(check: Check) -> None:
    assert check.title() == check._title  # noqa: SLF001
    assert check.description() == check._description  # noqa: SLF001
    assert check.hint() == check._hint  # noqa: SLF001

    result = await check.async_check()

    assert result.status() == check._status  # noqa: SLF001
    assert result.message() == check._message  # noqa: SLF001
    assert result.items() == check._items  # noqa: SLF001

    if check._error:  # noqa: SLF001
        assert isinstance(result.error(), openchecks.CheckError)

        assert str(result.error()) == str(check._error)  # noqa: SLF001
    else:
        assert result.error() is None

    if result.status() == openchecks.Status.SystemError:
        assert result.can_fix() is False
        assert result.can_skip() is False
    else:
        assert result.can_fix() == check._can_fix  # noqa: SLF001
        assert result.can_skip() == check._can_skip  # noqa: SLF001

    if result.status().has_failed() and result.can_fix():
        fix_result = await openchecks.async_auto_fix(check)

        if not check.hint() & openchecks.CheckHint.AUTO_FIX:
            assert fix_result.status() == openchecks.Status.SystemError
            assert fix_result.message() == "Check does not implement auto fix."
            assert fix_result.items() is None
            assert fix_result.error() is None
        elif fix_result.error():
            assert fix_result.status() == openchecks.Status.SystemError
            assert fix_result.message() == "Error in auto fix."
            assert fix_result.items() is None

            if check._error:  # noqa: SLF001
                assert isinstance(fix_result.error(), openchecks.CheckError)

                assert (
                    str(fix_result.error())
                    == f"{check._error.__class__.__name__}: {check._error}"  # noqa: SLF001
                )
            else:
                assert result.error() is None
        else:
            assert fix_result.status() == check._fix_status  # noqa: SLF001
            assert fix_result.message() == check._message  # noqa: SLF001
            assert fix_result.items() == check._items  # noqa: SLF001
            assert fix_result.error() is None

        if fix_result.status() == openchecks.Status.SystemError:
            assert fix_result.can_fix() is False
            assert fix_result.can_skip() is False
        else:
            assert fix_result.can_fix() == check._can_fix  # noqa: SLF001
            assert fix_result.can_skip() == check._can_skip  # noqa: SLF001


if __name__ == "__main__":
    import sys

    atheris.Setup(sys.argv, atheris.instrument_func(fuzz.hypothesis.fuzz_one_input))
    atheris.Fuzz()
